Build gives webpack error 98001. Generated engines use disallowed import "node-gyp-build"

### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Cloned from github.
gatsby new fg https://github.com/gatsbyjs/gatsby-starter-default
gatsby develop works fine.
gatsby build gives:
success load gatsby config - 0.023s
success load plugins - 0.641s
...
success Execute page configs - 0.029s
failed Validating Rendering Engines - 0.719s

 ERROR #98001  WEBPACK

Built Rendering Engines failed validation failed validation.

Please open an issue with a reproduction at
https://github.com/gatsbyjs/gatsby/issues/new for more help

  Error: Generated engines use disallowed import "node-gyp-build". Only allowed
  imports are to Node.js builtin modules or engines internals.

  - child.ts:60 Function._module.default._load
    [fg]/[gatsby]/src/utils/validate-engines/child.ts:60:11

  - loader:1005 Module.require
...

### Reproduction Link

https://github.com/gatsbyjs/gatsby-starter-default

### Steps to Reproduce

1. Clone the repository
2. gatsby build


### Expected Result

Expected the build to complete without error

### Actual Result

Build webpack error 98001

### Environment

```shell
System:
    OS: Windows 10 10.0.22000
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
  Binaries:
    Node: 16.14.0 - C:\Program Files\nodejs\node.EXE
    npm: 8.3.1 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.9.6 - /usr/bin/python
  Browsers:
    Edge: Spartan (44.22000.120.0), Chromium (98.0.1108.56)
  npmPackages:
    gatsby: ^4.8.0 => 4.8.0 
    gatsby-plugin-gatsby-cloud: ^4.8.0 => 4.8.0 
    gatsby-plugin-image: ^2.8.0 => 2.8.0 
    gatsby-plugin-manifest: ^4.8.0 => 4.8.0 
    gatsby-plugin-offline: ^5.8.0 => 5.8.0 
    gatsby-plugin-react-helmet: ^5.8.0 => 5.8.0 
    gatsby-plugin-sharp: ^4.8.0 => 4.8.0 
    gatsby-source-filesystem: ^4.8.0 => 4.8.0 
    gatsby-transformer-sharp: ^4.8.0 => 4.8.0 
  npmGlobalPackages:
    gatsby-cli: 4.8.0
```


### Config Flags

It is unchanged from the repo

This is a gatsby-plugin-sharp error, on Linux it says the below

Error: Something went wrong installing the "sharp" module
  Cannot find module '../build/Release/sharp-linux-x64.node'
  Require stack:
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/sharp/lib/sharp.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/sharp/lib/constructor.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/sharp/lib/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby-plugin-sharp/safe-sharp.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby-plugin-sharp/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby-transformer-sharp/customize-schema.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby-transformer-sharp/gatsby-node.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bootstrap/resolve-module-exports.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bootstrap/load-plugins/validate.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bootstrap/load-plugins/resolve-plugin.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bootstrap/load-themes/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bootstrap/load-config/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/services/initialize.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/services/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bootstrap/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/commands/build.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/node_modules/gatsby-cli/lib/create-cli.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/node_modules/gatsby-cli/lib/index.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/dist/bin/gatsby.js
  - /media/donald/D66CD5576CD5334B/Users/User1/Documents/Github/publiuslogic.co
  m/node_modules/gatsby/cli.js
  Possible solutions:
  - Install with the --verbose flag and look for errors: "npm install --ignore-
  scripts=false --verbose sharp"
  - Install for the current linux-x64 runtime: "npm install --platform=linux --
  arch=x64 sharp"
  - Consult the installation documentation: https://sharp.pixelplumbing.com/ins
  tall

I have had gatsby-image and now gatsby-plugin-image errors so many times since v 2, I am tired of this non-sense and am going to next which keeps it models stable and tests production installations and build on different platforms. Which obviously Gatsby does not, its almost like you are trying to break things. Which puts me and your other users through hell. Most of your packages have outdated modules of which some have been outdated for 5 to 6 years and there are alternatives other npm modules  to be used. Go through all plugins you have build and update them or you will keep seeing stupid errors with Gatsby and destroy your costumer base. I do not really find any new posts lately on Gatsby anything, all the best programmers in the last year are going to rust or next. Even a lot of the people who build and maintain Gatsby there personal websites are not Gatsby.

--- a/packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts
+++ b/packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts
@@ -71,7 +71,7 @@ export async function createGraphqlEngineBundle(
     module: {
       rules: [
         {
-          test: require.resolve(`lmdb`),
+          test: /node_modules[/\\]lmdb[/\\].*\.[cm]?js/,
           parser: { amd: false },
           use: [
             {
@@ -108,7 +108,7 @@ export async function createGraphqlEngineBundle(
         },
         {
           // For node binary relocations, include ".node" files as well here
-          test: /\.(m?js|node)$/,
+          test: /\.([cm]?js|node)$/,
           // it is recommended for Node builds to turn off AMD support
           parser: { amd: false },
           use: {
@@ -132,6 +132,8 @@ export async function createGraphqlEngineBundle(
         [require.resolve(`gatsby-cli/lib/reporter/loggers/ink/index.js`)]:
           false,
         inquirer: false,
+        // only load one version of lmdb
+        lmdb: require.resolve(`lmdb`),
       },
     },
     plugins: [

--- a/packages/gatsby/src/schema/graphql-engine/lmdb-bundling-patch.ts
+++ b/packages/gatsby/src/schema/graphql-engine/lmdb-bundling-patch.ts
@@ -1,4 +1,10 @@
-import { createRequireFromPath } from "gatsby-core-utils"
+/* eslint-disable @babel/no-invalid-this */
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+import { slash } from "gatsby-core-utils/path"
+import path from "path"
+
+// we need to require this module, we can't use import here
+const { createRequire } = require(`module`)
 
 // This is hacky webpack loader that does string replacements to
 // allow lmdb@2 to be bundled by webpack for engines.
@@ -16,20 +22,26 @@ import { createRequireFromPath } from "gatsby-core-utils"
 //  - https://github.com/DoctorEvidence/lmdb-js/blob/544b3fda402f24a70a0e946921e4c9134c5adf85/open.js#L77
 // Reliance on `import.meta.url` + usage of `.replace` is what seems to cause problems currently.
 
-export default function (source: string): string {
-  let lmdbBinaryLocation
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export default function (this: any, source: string): string {
+  let lmdbBinaryLocation: string | undefined
   try {
-    const lmdbRequire = createRequireFromPath(require.resolve(`lmdb`))
-    const nodeGypBuild = lmdbRequire(`node-gyp-build`)
-    const path = require(`path`)
+    const lmdbRoot =
+      this?._module.resourceResolveData?.descriptionFileRoot ||
+      path.dirname(this.resourcePath).replace(`/dist`, ``)
 
-    lmdbBinaryLocation = nodeGypBuild.path(
-      path.dirname(require.resolve(`lmdb`)).replace(`/dist`, ``)
+    const lmdbRequire = createRequire(this.resourcePath)
+    const nodeGypBuild = lmdbRequire(`node-gyp-build`)
+
+    lmdbBinaryLocation = slash(
+      path.relative(
+        path.dirname(this.resourcePath),
+        nodeGypBuild.path(lmdbRoot)
+      )
     )
   } catch (e) {
     return source
   }
-
   return source
     .replace(
       `require$1('node-gyp-build')(dirName)`,

